### PR TITLE
✨ Enable GitHub Enterprise Server (GHES) support

### LINKED ISSUE #660: Support private git servers (like Github Enterprise instances)
**Is your feature request related to a problem? Please describe.**
I would like to run scorecard for a repository that is hosted on my [Github Enterprise](https://enterprise.github.com/) instance. I currently can't.

```
$ scorecard --repo github.corp.net/nv35/myrepo
2021/07/05 09:24:35 unsupported host: github.corp.net
```

**Describe the solution you'd like**
Allow to use the others github instances just like the .com one. The API path would likely just change from https://api.github.com/ to https://github.corp.net/api/v3, and I guess the calls would just be the same.

**Additional context**
The restriction to github.com is defined here: https://github.com/ossf/scorecard/blob/1f1e05b22ccfac2c389125cdbc5be0bc61a52c1e/repos/repo_url.go#L77

### PR DESCRIPTION
**Pull request by rajbos**
#### What kind of change does this PR introduce?
Support GHES by using the Enterprise client for the GitHub calls.

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
``` shell
$ scorecard --repo github.corp.net/nv35/myrepo
unsupported host: github.corp.net
```

#### Which issue(s) this PR fixes
* Fixes #660

#### Special notes for your reviewer

There is one error that needs to be fixed, since this returns a non-zero exit code at the moment, hence the `Draft` status.

``` shell
Error: check runtime error: Branch-Protection: internal error: error during branchesHandler.setup: internal error: githubv4.Query: Field 'requireLastPushApproval' doesn't exist on type 'BranchProtectionRule'
2023/03/24 12:04:20 error during command execution: check runtime error: Branch-Protection: internal error: error during branchesHandler.setup: internal error: githubv4.Query: Field 'requireLastPushApproval' doesn't exist on type 'BranchProtectionRule'
exit status 1
``` 
I expect this is coming from our GHES instance (3.7) that does not support this field yet. If I remove the field [requireLastPushApproval](https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/clients/githubrepo/branches.go#L90) all checks run normally, so we might need to check the GHES version and include/exclude this field based on that.

#### Does this PR introduce a user-facing change?
Not for existing users, as it did not support GHES before. No changes needed by existing users.

## Discussion
I am using `os.Getenv("GITHUB_API_URL")` to pick up the API url for the server, as this is where I'd expect this to run from an automation context. We could pick up the url from the  `scorecard --repo github.corp.net/nv35/myrepo` command, but that means wiring up all the places from front to back to pick this up and pass it through and then update CreateGithubRepoClient calls everywhere.

```release-note
 Support for GitHub Enterprise Server has been added.
```

### CODE REVIEW

**Comment by rajbos** [comment]
I'm going to work on this again later this week. Not stale.

**Comment by azeemshaikh38** [comment]
@raghavkaul could you help review this PR whenever its ready?

**Comment by raghavkaul** [comment]
Yes, happy to review whenever this is ready.

**Comment by rajbos** [comment]
Tested against GHES 3.8.1 today and it works: the field in the GraphQL search exists. 

Two things left to discuss:
1. I need some help with the failing unit test: I have no clue how this works in Go and why this is failing (I get that I changed the setup partially)
2. Should this new change be reflected in the REAMDE (GHES is supported by looking at the `GITHUB_API_URL` variable. Is so: where? We can then note that this works from 3.8 and up.

**Comment by rajbos** [comment]
Tagging @raghavkaul for visiblity.

File: clients/githubrepo/client.go [hunk 1]
```diff
@@ -126,7 +128,7 @@ func (client *Client) InitRepo(inputRepo clients.Repo, commitSHA string, commitD
 
 // URI implements RepoClient.URI.
 func (client *Client) URI() string {
-	return fmt.Sprintf("github.com/%s/%s", client.repourl.owner, client.repourl.repo)
+	return fmt.Sprintf("%s/%s/%s", client.repourl.host, client.repourl.owner, client.repourl.repo)
```
Comment (naveensrinivasan) [inline clients/githubrepo/client.go:131]: Can you please write a test for this?

File: clients/githubrepo/client.go [hunk 2]
```diff
@@ -259,8 +261,33 @@ func CreateGithubRepoClientWithTransport(ctx context.Context, rt http.RoundTripp
 	httpClient := &http.Client{
 		Transport: rt,
 	}
-	client := github.NewClient(httpClient)
-	graphClient := githubv4.NewClient(httpClient)
+
```
Comment (naveensrinivasan) [inline clients/githubrepo/client.go:264]: Can you please include tests for these?

**Review by naveensrinivasan — COMMENTED** [review]
Thanks!

**Comment by rajbos** [comment]
Still want to work on this Stale bot! Looking for the time to do so :smile:

File: clients/githubrepo/client.go (hunk 2 above)
Comment (rajbos) [reply clients/githubrepo/client.go:264]: @naveensrinivasan , can you give me some pointers / help on where to get started? (not a Go programmer, so testing in Go is new to me). 

I see a generic setup for creating the clients during the tests here: [clients/githubrepo/githubrepo_suite_test.go](https://github.com/rajbos/scorecard-ghes/blob/b66365206d05232fe1b37ae24acbadc01733723f/clients/githubrepo/githubrepo_suite_test.go#L65), and another generic mock here [clients/mockclients/repo_client.go](https://github.com/rajbos/scorecard-ghes/blob/ghes-support/clients/mockclients/repo_client.go).

**Comment by esnible** [comment]
The error messages were not obvious to me:

```bash
go run main.go --repo https://github.ibm.com/<myorg>/<myproject>
2023/06/05 08:05:42 error during command execution: RunScorecard: repo unreachable: Get "/api/v3/repos/<myorg>/<myproject>": internal error: innerTransport.RoundTrip: internal error: innerTransport.RoundTrip: unsupported protocol scheme ""
```

Could this message remind people that you expect `$GITHUB_API_URL`, and the value it should have?

```bash
GITHUB_API_URL=https://api.github.ibm.com go run main.go --repo https://github.ibm.com/<myorg>/<myproject>
Error: RunScorecard: repo unreachable: GET https://api.github.ibm.com/repos/<myorg>/<myproject>: 401 Must authenticate to access this API. []
```

Could this message include the usual one `GitHub token env var is not set. Please read https://github.com/ossf/scorecard#authentication  error="an error occurred while getting GitHub credentials"`

```bash
GITHUB_AUTH_TOKEN=<mytoken> GITHUB_API_URL=https://api.github.ibm.com go run main.go --repo https://github.ibm.com/<myorg>/<myproject>
Error: RunScorecard: internal error: ListCommits:error during graphqlHandler.setup: internal error: githubv4.Query: non-200 OK status code: 404 Not Found body: "{\"message\":\"Not Found\",\"documentation_url\":\"https://docs.github.com/enterprise-server@3.3/rest\"}"
```

I gave up before solving this one.  Does it want the enterprise URL instead of docs.github.com?  How do I provide that?

**Comment by rajbos** [comment]
Missed the feedback from @esnible, will try to address that.

File: clients/githubrepo/client.go [hunk 3]
```diff
@@ -259,8 +261,33 @@ func CreateGithubRepoClientWithTransport(ctx context.Context, rt http.RoundTripp
...
+	if githubAPIURL := os.Getenv("GITHUB_API_URL"); githubAPIURL != "https://api.github.com" {
+		// create a new enterprise client with custom URLs
+		githubSERVERURL := os.Getenv("GITHUB_SERVER_URL")
+		if githubSERVERURL == "" {
+			// load the server url from the api url
+			githubSERVERURL = strings.TrimSuffix(githubAPIURL, "/api/v3")
+		}
+		// trim trailing slash to prevent issues with the graphql client
+		githubSERVERURL = strings.TrimSuffix(githubSERVERURL, "/")
+		githubGRAPHQLURL := fmt.Sprintf("%s/api/graphql", githubSERVERURL)
```
Comment (esnible) [inline clients/githubrepo/client.go:277]: This line didn't work for me.  To run the scorecards, I used `GITHUB_API_URL=https://api.github.ibm.com` and changed this line to

```golang
githubGRAPHQLURL := fmt.Sprintf("%s/graphql", githubSERVERURL)
```

Using /api/graphql I caused 404s, using /graphql allowed the scorecard to generate.  I only have access to my own company's GHE, so I don't know if you should always use /graphql or if this needs to be parameters for different configurations.

**Comment by spencerschrock** [comment]
FYI, https://github.com/ossf/scorecard/pull/2999 was merged while this was marked stale. 

Can we see if @esnible's feedback applies to what has already been merged? If not, can we re-purpose this PR to fixing what's already been merged?

**Comment by esnible** [comment]
Using the current HEAD I was able to `GITHUB_AUTH_TOKEN=$GITHUB_ENTERPRISE_TOKEN GH_HOST=github.ibm.com go run main.go --repo https://github.ibm.com/<owner>/<project> --show-details --checks Vulnerabilities` and it worked just fine.

I had incorrectly tried `GH_HOST=api.github.ibm.com` with HEAD, as that is the host I use when I make GraphQL queries, and that is the host that this PR expected, and I found the error message confusing.  However, I was not confused enough to be stuck.  The error message was:

```
error during command execution: GetClients: error making github repo: error in IsValid: unsupported host: github.ibm.com
```

That confused me because
- I wasn't making a github repo
- I was passing in a GH_HOST that I knew was supported
- I thought the error message referred to $GH_HOST

In terms of functionality the code in HEAD is great.

**Comment by spencerschrock** [comment]
> I've validated with [v4.11.0](https://github.com/ossf/scorecard/releases/tag/v4.11.0) and it works against our internal GHES environment, so I am ready to close this issue. Thanks for adding the support!

All the thanks goes to you and Niket!

**Pull request closed without merging**